cocotb/utils.py

```diff
@@ -202,97 +202,127 @@ def unpack(ctypes_obj, string, bytes=None):
 import cocotb.ANSI as ANSI
 
 
-def _sane_color(x):
+# A note on the use of latin1 in the deprecations below:
+# Latin1 is the only encoding `e` that satisfies
+# `all(chr(x).encode(e) == bytes([x]) for x in range(255))`
+# Our use of `ord` and `chr` throughout other bits of code make this the most
+# compatible choice of encoding. Under this convention, old code can be upgraded
+# by changing "binary" strings from `"\x12\x34"` to `b"\x12\x34"`.
+
+
+def _sane_color(x: bytes) -> str:
     r = ""
-    for i in x:
-        j = ord(i)
+    for j in x:
         if (j < 32) or (j >= 127):
             r += "."
         else:
-            r += i
+            r += chr(j)
     return r
 
 
-def hexdump(x):
+def hexdump(x: bytes) -> str:
     """Hexdump a buffer.
 
     Args:
-        x: Object that supports conversion via the ``str`` built-in.
+        x: Object that supports conversion to :class:`bytes`.
 
     Returns:
         A string containing the hexdump.
 
+    .. deprecated:: 1.4
+        Passing a :class:`str` to this function is deprecated, as it
+        is not an appropriate type for binary data. Doing so anyway
+        will encode the string to latin1.
+
     Example:
-        >>> print(hexdump('this somewhat long string'))
+        >>> print(hexdump(b'this somewhat long string'))
         0000   74 68 69 73 20 73 6F 6D 65 77 68 61 74 20 6C 6F   this somewhat lo
         0010   6E 67 20 73 74 72 69 6E 67                        ng string
         <BLANKLINE>
     """
     # adapted from scapy.utils.hexdump
     rs = ""
-    x = str(x)
+    if isinstance(x, str):
+        warnings.warn(
+            "Passing a string to hexdump is deprecated, pass bytes instead",
+            DeprecationWarning, stacklevel=2)
+        x = x.encode('latin1')
+    x = b"%b" % x
     l = len(x)
     i = 0
     while i < l:
         rs += "%04x   " % i
         for j in range(16):
             if i + j < l:
-                rs += "%02X " % ord(x[i + j])
+                rs += "%02X " % x[i + j]
             else:
                 rs += "   "
             if j % 16 == 7:
                 rs += ""
         rs += "  "
+        # latin1 is fine, because we conver
         rs += _sane_color(x[i:i + 16]) + "\n"
         i += 16
     return rs
 
 
-def hexdiffs(x, y):
-    """Return a diff string showing differences between two binary strings.
+def hexdiffs(x: bytes, y: bytes) -> str:
```

Comment:
Wondering why this is called ``hexdiffs`` and not ``hexdiff``(as in scapy.utils), and whether we should rename this to ``hexdiff`` and make ``hexdiffs`` an alias (that can be deprecated)?

Reply:
Not in this PR maybe.

Reply:
Not in this PR, I agree